### PR TITLE
Optimization for batched processing

### PR DESCRIPTION
**Pull request by HYLcool**
- Extract the batched processing loop to a new function (e.g. compute_stats_batched), and decide the function branch to be called when processing data.
- Pros:
  - No need to modify each OP to support batched processing. Only need to set the arg `_batched_op` to True for each OP.
  - Developers can add new OPs in a single-sample-dict way, which is more friendly.
- Cons (for now):
  - Slightly slower data processing: 4 scenarios were tested and the results are shown below:

| Scenario                    | OPs                                      | Before   | After             |
|-----------------------------|------------------------------------------|----------|-------------------|
| Simple Filter               | `text_length_filter`                       | 4.850s   | 5.340s (+10.10%)  |
| Simple Mapper               | `whitespace_normalization_mapper`          | 12.145s  | 12.575s (+3.54%)  |
| With `context` (OP fusion)    | `word_num_filter` & `word_repetition_filter` | 80.456s  | 87.315s (+8.53%)  |
| With `rank` (Model-based OPs) | `image_text_similarity_filter`             | 562.586s | 590.502s (+4.96%) |

### CODE REVIEW

File: data_juicer/ops/base_op.py [hunk 1]
```diff
@@ -236,9 +236,21 @@ def __init__(self, *args, **kwargs):
...
-            self.process = catch_map_batches_exception(self.process)
+            self.process_branch = catch_map_batches_exception(
+                self.process_batched)
         else:
-            self.process = catch_map_single_exception(self.process)
+            self.process_branch = catch_map_single_exception(self.process)
+
+    def process_batched(self, samples, *args, **kwargs):
+        keys = samples.keys()
+        first_key = list(keys)[0]
```
Comment (drcege) [inline data_juicer/ops/base_op.py:246]: 建议
```python
first_key  = next(iter(keys))
num_samples = len(samples[first_key])
```

File: data_juicer/ops/base_op.py [hunk 2]
```diff
@@ -284,10 +296,30 @@ def __init__(self, *args, **kwargs):
...
         else:
-            self.compute_stats = catch_map_single_exception(self.compute_stats)
+            self.compute_stats_branch = catch_map_single_exception(
+                self.compute_stats)
+            self.process_branch = catch_map_single_exception(self.process)
+
+    def compute_stats_batched(self, samples, *args, **kwargs):
+        keys = samples.keys()
+        samples_stats = samples[Fields.stats]
+        for i in range(len(samples_stats)):
```
Comment (drcege) [inline data_juicer/ops/base_op.py:311]: 最好其他 OP 都统一成相同的模式和变量名，比如
```python
num_samples = len(samples[Fields.stats])
for i in range(num_samples):
```

File: data_juicer/ops/base_op.py [hunk 3]
```diff
@@ -284,10 +296,30 @@ def __init__(self, *args, **kwargs):
...
+            this_sample = {key: samples[key][i] for key in keys}
+            res_sample = self.compute_stats(this_sample, *args, **kwargs)
+            samples[Fields.stats][i] = res_sample[Fields.stats]
+            if 'context' in kwargs and kwargs['context']:
+                samples[Fields.context][i] = res_sample[Fields.context]
+
+        return samples
+
+    def process_batched(self, samples):
+        return map(lambda stat: self.process({Fields.stats: stat}),
```
Comment (drcege) [inline data_juicer/ops/base_op.py:321]: 确认 `ray` 下是否正确

File: data_juicer/ops/base_op.py [hunk 4]
```diff
@@ -236,9 +236,21 @@ def __init__(self, *args, **kwargs):
 
         # runtime wrappers
         if self.is_batched_op():
-            self.process = catch_map_batches_exception(self.process)
+            self.process_branch = catch_map_batches_exception(
```
Comment (drcege) [inline data_juicer/ops/base_op.py:239]: `_branch` 是什么含义？比如 `self.process_safe` 或类似

Comment (HYLcool) [reply data_juicer/ops/base_op.py:239]: 还是改回了原来的compute_stats和process，但是是否batch化的两个函数分别重命名成为了带"_batched"和"_single"后缀的版本，例如：
```Python
        if self.is_batched_op():
            self.process = catch_map_batches_exception(self.process_batched)
        else:
            self.process = catch_map_single_exception(self.process_single)
```

File: data_juicer/ops/base_op.py (hunk 3 above)
Comment (HYLcool) [reply data_juicer/ops/base_op.py]: @pan-x-c

Comment (HYLcool) [reply data_juicer/ops/base_op.py]: 感谢@pan-x-c 帮忙验证，ray那边也是ok的👌🏻

File: data_juicer/ops/base_op.py [hunk 5]
```diff
@@ -236,11 +236,23 @@ def __init__(self, *args, **kwargs):
 
         # runtime wrappers
         if self.is_batched_op():
-            self.process = catch_map_batches_exception(self.process)
+            self.process = catch_map_batches_exception(self.process_batched)
```
Comment (drcege) [inline data_juicer/ops/base_op.py:239]: 需要限制 `self.process` 不能被子类重写吗，因为可能与变更前的接口混淆。

比如，严格阻止重写：
```python
class Base:
    def __init_subclass__(cls, **kwargs):
        if 'my_method' in cls.__dict__:
            raise TypeError(f"{cls.__name__} cannot override my_method")

    def my_method(self):
        print("This method cannot be overridden.")

class Derived(Base):
    def my_method(self):  # 这将引发 TypeError
        print("Trying to override.")
```

或者至少静态分析警告：
```python
from typing import final

class Base:
    @final
    def my_method(self):
        print("This method cannot be overridden.")

class Derived(Base):
    # 尝试重写 my_method 会导致警告，但程序仍然可以运行
    def my_method(self):
        print("Trying to override.")
```

File: docs/DeveloperGuide_ZH.md [hunk 6]
```diff
@@ -116,14 +117,15 @@ class StatsKeys(object):
...
+        def compute_stats_single(self, sample, rank=None):
             # ... (same as above)
 
-        def process(self, sample, rank=None):
+        def process_single(self, sample, rank=None):
             # ... (same as above)
     ```
 
-    - 如果算子批量处理数据，输入不是一个样本而是一个batch，需要声明`_batched_op = True`。
+    - 如果算子批量处理数据，输入不是一个样本而是一个batch，或者你想在单样本实现上直接激活batch化处理，需要声明`_batched_op = True`。
```
Comment (drcege) [inline docs/DeveloperGuide_ZH.md:127]: 下一步考虑去掉 `_batched_op`，通过给算子预设不同 `batch_size` 来控制 batch 与否。

**Comment by drcege** [comment]
Please sync the main branch.

File: data_juicer/ops/base_op.py (hunk 5 above)
Comment (HYLcool) [reply data_juicer/ops/base_op.py:239]: 好主意~采用了第一种方法👍🏻

File: docs/DeveloperGuide_ZH.md (hunk 6 above)
Comment (HYLcool) [reply docs/DeveloperGuide_ZH.md:127]: 这个部分需要额外考虑不少其他参数，与当前PR不是强绑定，也许可以在下一个PR里进行

File: data_juicer/ops/base_op.py (hunk 5 above)
Comment (drcege) [reply data_juicer/ops/base_op.py:239]: 或者只是简单重命名呢😄，`self._process`，你评估下

Comment (HYLcool) [reply data_juicer/ops/base_op.py:239]: 我感觉强制要求统一用法好一些，也就是你给的第一种方案

**Review by drcege — APPROVED** [review]
LGTM

**Pull request merged**